File: 二叉树/节点与其祖先之间的最大差值.java
package 二叉树;

public class 节点与其祖先之间的最大差值 {
    int res;

    public int maxAncestorDiff(TreeNode root) {
        dfs(root, root.val, root.val);
        return res;
    }

    //每层都必须维护一个该节点独有的最大值最小值，所以max 、 min作为局部变量传递
    public void dfs(TreeNode root, int max, int min) {
        if (root == null) return;
        res = Math.max(Math.abs(max - root.val), Math.max(res, Math.abs(min - root.val)));
        if (root.val > max) {
            max = root.val;
        }
        if (root.val < min) {
            min = root.val;
        }
        dfs(root.left, max, min);
        dfs(root.right, max, min);
    }
}

File: 二叉树/节点与其祖先之间的最大差值.c

#include <stddef.h>
#include <stdlib.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define MIN(a, b) ((a) < (b) ? (a): (b))


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int res;

void dfs(struct TreeNode *root, int min, int max);

int maxAncestorDiff(struct TreeNode *root) {
    res = 0;
    dfs(root, root->val, root->val);
    return res;
}

void dfs(struct TreeNode *root, int min, int max) {
    if (root == NULL) return;
    res = MAX(res, abs(root->val - min));
    res = MAX(res, abs(root->val - max));
    min = MIN(min, root->val);
    max = MAX(max, root->val);
    dfs(root->left, min, max);
    dfs(root->right, min, max);
}


